src/Form/MediaLibraryEmbeddableForm.php

```diff
@@ -0,0 +1,163 @@
+
+/**
+ * Media library add stanford embed input form.
+ *
+ * @package Drupal\stanford_media\Form
+ */
+class MediaLibraryEmbeddableForm extends OEmbedForm {
+
+  /**
+   * {@inheritDoc}
+   */
+  public function getFormId() {
+    return $this->getBaseFormId() . '_embeddable';
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  protected function buildInputElement(array $form, FormStateInterface $form_state) {
+    // This was adapted from \Drupal\media_library\Form\OembedForm.
+    $user = \Drupal::currentUser();
+    $authorized_for_unstructured = $user->hasPermission('create field_media_embeddable_code') || $user->hasPermission('edit field_media_embeddable_code');
+
+    $media_type = $this->getMediaType($form_state);
+    $providers = $media_type->getSource()->getProviders();
+
+    $form['container'] = [
+      '#type' => 'container',
+    ];
+
+    $form['container']['field_media_embeddable_oembed'] = [
+      '#type' => 'url',
+      '#title' => $this->t('Add @type via URL', [
+        '@type' => $this->getMediaType($form_state)->label(),
+      ]),
+      '#description' => $this->t('Allowed providers: @providers. For custom embeds, please request support.', [
+        '@providers' => implode(', ', $providers),
+      ]),
+      '#required' => FALSE,
+      '#attributes' => [
+        'placeholder' => 'https://',
+      ],
+    ];
+
+    if ($authorized_for_unstructured) {
+      $form['container']['field_media_embeddable_code'] = [
+        '#type' => 'textarea',
+        '#title' => $this->t('Embed Code'),
+        '#description' => $this->t('Use this field to paste in embed codes which are not available through oEmbed'),
+        '#required' => FALSE,
+        '#attributes' => [
+          'placeholder' => '',
+        ],
+      ];
+    }
+
+    $ajax_query = $this->getMediaLibraryState($form_state)->all();
+    $ajax_query += [FormBuilderInterface::AJAX_FORM_REQUEST => TRUE];
+    $form['container']['submit'] = [
+      '#type' => 'submit',
+      '#value' => $this->t('Add'),
+      '#button_type' => 'primary',
+      '#validate' => ['::validateEmbeddable'],
+      '#submit' => ['::addButtonSubmit'],
+      '#ajax' => [
+        'callback' => '::updateFormCallback',
+        'wrapper' => 'media-library-wrapper',
+        // Add a fixed URL to post the form since AJAX forms are automatically
+        // posted to <current> instead of $form['#action'].
+        // @todo Remove when https://www.drupal.org/project/drupal/issues/2504115
+        //   is fixed.
+        'url' => Url::fromRoute('media_library.ui'),
+        'options' => ['query' => $ajax_query],
+      ],
+    ];
+
+    return $form;
+  }
+
+  /**
+   * Informs us if we are working with an unstructured embed.
+   *
+   * @param \Drupal\Core\Form\FormStateInterface $form_state
+   *   The current form state.
+   * @return bool
+   *   True if unstructured, otherwise false.
+   */
+  public function isUnstructured(FormStateInterface $form_state) {
```

Comment:
If this is only used internally, it is best to make it a protected method instead of public.

Reply:
I see it is public. Nevermind then.